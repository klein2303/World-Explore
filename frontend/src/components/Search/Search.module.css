.searchcontainer {
    border: 2px solid #e4e1f7;
    height: fit-content;
    width: 500px;
    border-radius: 10px;

    display: flex;
    flex-direction: row;
    justify-content: space-between;
    padding: 5px;
}

.input {
    border: none;
    width: 460px;
    color: #424242;
}

.input:focus {
    outline: none;
}

.searchicon {
    background-color: #e4e1f7;
    color: #424242;
    font-size: 20px;
    border-radius: 20px;
    padding: 4px;

    cursor: pointer;
}

@media (max-width: 950px) {
    .searchcontainer {
        width: 330px;
        padding: 2px;
    }

    .input {
        width: 300px;
    }
}

@media (max-width: 800px) {
    .searchcontainer {
        width: 300px;
        padding: 2px;
    }

    .input {
        width: 270px;
    }
}

@media (max-width: 620px) {
    .searchcontainer {
        width: 280px;
        padding: 2px;
    }

    .searchicon {
        padding: 3px;
        border-radius: 15px;
        font-size: 16px;
    }
    .input {
        width: 250px;
    }
}
