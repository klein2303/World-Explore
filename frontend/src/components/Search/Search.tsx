import styles from "./Search.module.css";
import {GoSearch} from "react-icons/go";
import { filterAtom } from "../../atoms/FilterAtom";
import { useRecoilState } from "recoil";
import { useEffect, useState} from "react";
import { FilterType } from "../../types/FilterType";
import { setFilters, getFilters} from "../../utils/FilterStorage";
import { useDebounce } from 'use-debounce';


const Search = () => {
    const [filter, setFilter] = useRecoilState<FilterType>(filterAtom);
    const [input, setInput] = useState<string>(getFilters().search);
    const [debouncedSearch] = useDebounce(input, 1000);

    const handleInputChange = (e: React.ChangeEvent<HTMLInputElement>) => {
        setInput(e.target.value);  
    }

    useEffect(() => {
        if (debouncedSearch == filter.search) return;
        setFilter((prevState) => ({
            ...prevState,
            search: debouncedSearch,
        }));
        setFilters(filter);
    }, [debouncedSearch, setFilter, filter]);

    return (
        <main className={styles.searchcontainer}>
            <input type="text" placeholder="Search after countries..." className={styles.input} aria-label="Search countries" onChange={handleInputChange} value={input}/>
            <GoSearch className={styles.searchicon}  role="button" aria-label="Click to search" tabIndex={0}/>
        </main>
    );
};

export default Search;