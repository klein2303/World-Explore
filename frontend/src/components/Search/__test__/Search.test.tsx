import { render, screen, fireEvent } from "@testing-library/react";
import "@testing-library/jest-dom";
import { describe, it, expect } from "vitest";
import { MemoryRouter } from "react-router-dom";
import Search from "../Search";

describe("Search Component Tests", () => {
    it("renders the search input and icon correctly", () => {
        // Render the Search component 
        const { asFragment } = render(
            <MemoryRouter>
                <Search />
            </MemoryRouter>
        );

        // Take a snapshot of the initial render
        expect(asFragment()).toMatchSnapshot();

        // Check if the search input field is present
        const inputElement = screen.getByPlaceholderText("Search after countries...");
        expect(inputElement).toBeInTheDocument();

        // Check if the search button (icon) is present
        const searchButton = screen.getByRole("button", { name: "Click to search" });
        expect(searchButton).toBeInTheDocument();
    });

    it("allows typing in the search input", () => {
        // Render the Search component
        render(
            <MemoryRouter>
                <Search />
            </MemoryRouter>
        );

        // Get the input element by its placeholder text
        const inputElement = screen.getByPlaceholderText("Search after countries...");

        // Simulate typing in the input field
        fireEvent.change(inputElement, { target: { value: "Canada" } });

        // Check if the input value has been updated
        expect(inputElement).toHaveValue("Canada");
    });
});