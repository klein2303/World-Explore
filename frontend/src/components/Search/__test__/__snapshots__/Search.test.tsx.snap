// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Search Component Tests > renders the search input and icon correctly 1`] = `
<DocumentFragment>
  <main
    class="_searchcontainer_a9d8a5"
  >
    <input
      aria-label="Search countries"
      class="_input_a9d8a5"
      placeholder="Search after countries..."
      type="text"
    />
    <svg
      aria-label="Click to search"
      class="_searchicon_a9d8a5"
      fill="currentColor"
      height="1em"
      role="button"
      stroke="currentColor"
      stroke-width="0"
      tabindex="0"
      viewBox="0 0 24 24"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M10.25 2a8.25 8.25 0 0 1 6.34 13.53l5.69 5.69a.749.749 0 0 1-.326 1.275.749.749 0 0 1-.734-.215l-5.69-5.69A8.25 8.25 0 1 1 10.25 2ZM3.5 10.25a6.75 6.75 0 1 0 13.5 0 6.75 6.75 0 0 0-13.5 0Z"
      />
    </svg>
  </main>
</DocumentFragment>
`;
